### ISSUE TITLE
i586: /usr/include/gnu/stubs.h:10:11: fatal error: 'gnu/stubs-64.h' file not found

### ISSUE THREAD

**Issue by vt-alt** [ispc/ispc#2066]
I try to build `ispc` package for ALT Linux for 32-bit x86. There is a compilation error:
```
builder@i586:~/RPM/BUILD/ispc-1.15.0$ make -C BUILD VERBOSE=1
...
[  2%] Generating builtins-cpp-64-linux-x86_64.cpp
cd /usr/src/RPM/BUILD/ispc-1.15.0 && /usr/lib/llvm-11.0/bin/clang++ --target=x86_64-unknown-linux-gnu -fPIC -isystem/usr/x86_64-linux-gnu/include -I/usr/src/RPM/BUILD/ispc-1.15.0 -m64 -emit-llvm --std=gnu++17 -c builtins/builtins-c-cpu.cpp -o - | ( "/usr/lib/llvm-11.0/bin/llvm-dis" - || echo builtins-c-*.cpp\ compile\ error ) | "/usr/bin/python3.9" bitcode2cpp.py c --type=builtins-c --runtime=64 --os=linux --arch=x86_64 --llvm_as /usr/lib/llvm-11.0/bin/llvm-as > /usr/src/RPM/BUILD/ispc-1.15.0/BUILD/./builtins-cpp-64-linux-x86_64.cpp
In file included from builtins/builtins-c-cpu.cpp:69:
In file included from /usr/include/unistd.h:25:
In file included from /usr/include/features.h:489:
/usr/include/gnu/stubs.h:10:11: fatal error: 'gnu/stubs-64.h' file not found
# include <gnu/stubs-64.h>
          ^~~~~~~~~~~~~~~~
1 error generated.
/usr/lib/llvm-11.0/bin/llvm-dis: error: file too small to contain bitcode header
/usr/lib/llvm-11.0/bin/llvm-as: <stdin>:1:1: error: expected top-level entity
builtins-c-*.cpp compile error
^
make[2]: *** [CMakeFiles/ispc.dir/build.make:342: builtins-cpp-64-linux-x86_64.cpp] Error 1
make[2]: *** Deleting file 'builtins-cpp-64-linux-x86_64.cpp'
make[2]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make[1]: *** [CMakeFiles/Makefile2:153: CMakeFiles/ispc.dir/all] Error 2
make[1]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make: *** [Makefile:149: all] Error 2
make: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
```
We don't have `stubs-64.h` on i586 platform.
```
builder@i586:~/RPM/BUILD/ispc-1.15.0$ ls -l /usr/include/gnu/stubs*
-rw-r--r-- 1 rooter rooter 471 Dec 18 11:20 /usr/include/gnu/stubs-32.h
-rw-r--r-- 1 rooter rooter 384 Dec 18 11:20 /usr/include/gnu/stubs.h
```

I also read this https://github.com/ispc/ispc/issues/541#issuecomment-21056493

But, I still wonder why it wants to compile 64-bit binary for i586.

`sed -i s/-m64/-m32/g ./BUILD/CMakeFiles/ispc.dir/build.make` helps. Maybe there should option for this?

**Comment by vt-alt** [comment]
This trick also works (to workaround this error):
```
    6  mkdir gnu
    7  cp /usr/include/gnu/stubs-32.h gnu/stubs-64.h
    8  make -C BUILD
```
But, I'm not sure it won't cause some unexpected consequences.

**Comment by vt-alt** [comment]
There is a similar problem on arm32:
```
builder@armh:~/RPM/BUILD/ispc-1.15.0$ make -j64 -C BUILD VERBOSE=1
...
[  1%] Generating builtins-cpp-64-linux-aarch64.cpp
make[2]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make  -f CMakeFiles/check_isa.dir/build.make CMakeFiles/check_isa.dir/build
cd /usr/src/RPM/BUILD/ispc-1.15.0 && /usr/lib/llvm-11.0/bin/clang++ --target=aarch64-unknown-linux-gnu -fPIC -isystem/usr/aarch64-linux-gnu/include -I/usr/src/RPM/BUILD/ispc-1.15.0 -m64 -emit-llvm --std=gnu++17 -c builtins/builtins-c-cpu.cpp -o - | ( "/usr/lib/llvm-11.0/bin/llvm-dis" - || echo builtins-c-*.cpp\ compile\ error ) | "/usr/bin/python3.9" bitcode2cpp.py c --type=builtins-c --runtime=64 --os=linux --arch=aarch64 --llvm_as /usr/lib/llvm-11.0/bin/llvm-as > /usr/src/RPM/BUILD/ispc-1.15.0/BUILD/./builtins-cpp-64-linux-aarch64.cpp
make[2]: Entering directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make[2]: Nothing to be done for 'CMakeFiles/check_isa.dir/build'.
make[2]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
[  5%] Built target check_isa
In file included from builtins/builtins-c-cpu.cpp:69:
In file included from /usr/include/unistd.h:25:
In file included from /usr/include/features.h:489:
/usr/include/gnu/stubs.h:7:11: fatal error: 'gnu/stubs-soft.h' file not found
# include <gnu/stubs-soft.h>
          ^~~~~~~~~~~~~~~~~~
1 error generated.
/usr/lib/llvm-11.0/bin/llvm-dis: error: file too small to contain bitcode header
/usr/lib/llvm-11.0/bin/llvm-as: <stdin>:1:1: error: expected top-level entity
builtins-c-*.cpp compile error
^
make[2]: *** [CMakeFiles/ispc.dir/build.make:137: builtins-cpp-64-linux-aarch64.cpp] Error 1
make[2]: *** Deleting file 'builtins-cpp-64-linux-aarch64.cpp'
make[2]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make[1]: *** [CMakeFiles/Makefile2:153: CMakeFiles/ispc.dir/all] Error 2
make[1]: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
make: *** [Makefile:149: all] Error 2
make: Leaving directory '/usr/src/RPM/BUILD/ispc-1.15.0/BUILD'
```
Replacing `-m64` with `-m32` (as worked with i586) does not help. It adds warning:
```
clang-11: warning: unknown platform, assuming -mfloat-abi=soft
```
And repeats the same error.

**Comment by dbabokin** [comment]
What cmake configuration was used? Do you need a cross build?

**Comment by vt-alt** [comment]
This is native builds. We build using (in rpm spec):
```
%cmake \
        -DCMAKE_BUILD_TYPE=RelWithDebInfo \
        -DCMAKE_INSTALL_PREFIX=%prefix \
        -DCMAKE_EXE_LINKER_FLAGS="%optflags -fPIE" \
        -DLLVM_DIR=$(llvm-config --cmakedir) \
        -DISPC_NO_DUMPS=ON \
        -DISPC_INCLUDE_TESTS=OFF \
        -DISPC_INCLUDE_EXAMPLES=OFF \
```
Which expands to: `cmake .. -DCMAKE_SKIP_RPATH:BOOL=ON -DCMAKE_SKIP_INSTALL_RPATH:BOOL=yes '-DCMAKE_C_FLAGS:STRING=-pipe -frecord-gcc-switches -Wall -g -O2 -march=i586 -mtune=generic' '-DCMAKE_CXX_FLAGS:STRING=-pipe -frecord-gcc-switches -Wall -g -O2 -march=i586 -mtune=generic' '-DCMAKE_Fortran_FLAGS:STRING=-pipe -frecord-gcc-switches -Wall -g -O2 -march=i586 -mtune=generic' -DCMAKE_INSTALL_PREFIX=/usr -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DLIB_DESTINATION=lib -DLIB_SUFFIX= -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_INSTALL_PREFIX=/usr '-DCMAKE_EXE_LINKER_FLAGS=-pipe -frecord-gcc-switches -Wall -g -O2 -march=i586 -mtune=generic -fPIE' -DLLVM_DIR=/usr/lib/llvm-11.0/lib/cmake/llvm -DISPC_NO_DUMPS=ON -DISPC_INCLUDE_TESTS=OFF -DISPC_INCLUDE_EXAMPLES=OFF`.

**Comment by dbabokin** [comment]
I'm trying to reproduce the issue in docker. What version should I use? `alt:latest` or `alt:p9`?

We also recommend building LLVM as part of ISPC build instead of using system LLVM. Is it possible in the framework of ALT Linux packages?

**Comment by vt-alt** [comment]
It should be `alt:latest`. Thanks. You would need to install `apt-get install rpm-build` to get a basic build environment. (Besides all requirements from `BuildRequires` from the spec.)

> We also recommend building LLVM as part of ISPC build instead of using system LLVM. Is it possible in the framework of ALT Linux packages?

This defies the packaging system, so would be highly discouraged. Generally, I can build ispc using system llvm with these workarounds.

(What I currently don't understand is why to create `-m64` binary for a 32-bit system.)

**Comment by vt-alt** [comment]
ps. Also, we try to have the latest versions of packages for our 'latest' (Sisyphus) branch, so Clang/LLVM should not be too old like in other distributions. Currently, we have LLVM 12.0.

**Comment by pbrubaker** [comment]
Closing this issue as inactive.  Please reopen if still relevant.

**Comment by LAKostis** [comment]
Just tested v1.25.3 with llvm 19.1.7 on ALTLinux i586 build, still fails with the same error:

[  2%] Generating share/ispc/builtins_dispatch.bc
cd /usr/src/RPM/BUILD/ispc-1.25.3 && /usr/bin/clang++ -x c -DREGULAR -fPIC --target=x86_64-unknown-unknown -march=core2 -mtune=generic -O2 -emit-llvm builtins/dispatch.c -c -o /usr/src/RPM/BUILD/isp
In file included from builtins/dispatch.c:24:
In file included from /usr/lib/llvm-19.1/lib/clang/19/include/stdint.h:56:
In file included from /usr/include/stdint.h:26:
In file included from /usr/include/bits/libc-header-start.h:33:
In file included from /usr/include/features.h:535:
/usr/include/gnu/stubs.h:10:11: fatal error: 'gnu/stubs-64.h' file not found
   10 | # include <gnu/stubs-64.h>
      |           ^~~~~~~~~~~~~~~~
1 error generated.

I think the problem comes from the fact how host CPU is detected, we use i386 chroot with 64-bit host and cpuid returns x86_64 instead of x86.

**Comment by nurmukhametov** [comment]
> Just tested v1.25.3 with llvm 19.1.7 on ALTLinux i586 build, still fails with the same error:
> 
> [ 2%] Generating share/ispc/builtins_dispatch.bc cd /usr/src/RPM/BUILD/ispc-1.25.3 && /usr/bin/clang++ -x c -DREGULAR -fPIC --target=x86_64-unknown-unknown -march=core2 -mtune=generic -O2 -emit-llvm builtins/dispatch.c -c -o /usr/src/RPM/BUILD/isp In file included from builtins/dispatch.c:24: In file included from /usr/lib/llvm-19.1/lib/clang/19/include/stdint.h:56: In file included from /usr/include/stdint.h:26: In file included from /usr/include/bits/libc-header-start.h:33: In file included from /usr/include/features.h:535: /usr/include/gnu/stubs.h:10:11: fatal error: 'gnu/stubs-64.h' file not found 10 | # include <gnu/stubs-64.h> | ^~~~~~~~~~~~~~~~ 1 error generated.
> 
> I think the problem comes from the fact how host CPU is detected, we use i386 chroot with 64-bit host and cpuid returns x86_64 instead of x86.

Could you share the CMake command you are using to build ISPC?

**Comment by nurmukhametov** [comment]
> > Just tested v1.25.3 with llvm 19.1.7 on ALTLinux i586 build, still fails with the same error:
> > [ 2%] Generating share/ispc/builtins_dispatch.bc cd /usr/src/RPM/BUILD/ispc-1.25.3 && /usr/bin/clang++ -x c -DREGULAR -fPIC --target=x86_64-unknown-unknown -march=core2 -mtune=generic -O2 -emit-llvm builtins/dispatch.c -c -o /usr/src/RPM/BUILD/isp In file included from builtins/dispatch.c:24: In file included from /usr/lib/llvm-19.1/lib/clang/19/include/stdint.h:56: In file included from /usr/include/stdint.h:26: In file included from /usr/include/bits/libc-header-start.h:33: In file included from /usr/include/features.h:535: /usr/include/gnu/stubs.h:10:11: fatal error: 'gnu/stubs-64.h' file not found 10 | # include <gnu/stubs-64.h> | ^~~~~~~~~~~~~~~~ 1 error generated.
> > I think the problem comes from the fact how host CPU is detected, we use i386 chroot with 64-bit host and cpuid returns x86_64 instead of x86.

This compilation failure is reproduced on v1.25.3 but not on v1.26.0. Try the newer version.

**Comment by LAKostis** [comment]
yes, I can confirm, no more failures with v1.26.0. Build passed together with tests.
